src/libqof/qof/qofsession.cpp

```diff
@@ -656,237 +449,256 @@ qof_session_save (QofSession *session,
+                m_saving = false;
+                return;
+            }
         }
-
         /* If we got to here, then the backend saved everything
         * just fine, and we are done. So return. */
-        /* Return the book_id to previous value. */
-        qof_session_clear_error (session);
+        clear_error ();
         LEAVE("Success");
-        goto leave;
+        m_saving = false;
+        return;
     }
     else
     {
-        if (ERR_BACKEND_NO_ERR != qof_session_get_error(session))
+        if (ERR_BACKEND_NO_ERR != get_error ())
         {
-	    /* push_error strdups, stack const is fine. */
-	    const char *msg = "failed to load backend";
-            qof_session_push_error(session, ERR_BACKEND_NO_HANDLER, msg);
+            std::string msg {"failed to load backend"};
+            push_error (ERR_BACKEND_NO_HANDLER, msg);
         }
     }
     LEAVE("error -- No backend!");
-leave:
-    g_atomic_int_inc(&session->lock);
-    return;
+    m_saving = false;
 }
 
 void
-qof_session_safe_save(QofSession *session, QofPercentageFunc percentage_func)
+QofSessionImpl::safe_save (QofPercentageFunc percentage_func) noexcept
 {
-    QofBackend *be = session->backend;
-    QofBackendError err;
-    char *msg = NULL;
-    g_return_if_fail( be != NULL );
-    g_return_if_fail( be->safe_sync != NULL );
+    auto be = m_backend;
+    if (!be) return;
+    if (!be->safe_sync) return;
     be->percentage = percentage_func;
-    (be->safe_sync)( be, qof_session_get_book( session ));
-    err = qof_backend_get_error(session->backend);
-    msg = qof_backend_get_message(session->backend);
+    (be->safe_sync) (be, get_book ());
+    auto err = qof_backend_get_error (m_backend);
+    auto msg = qof_backend_get_message (m_backend);
     if (err != ERR_BACKEND_NO_ERR)
     {
-        g_free(session->book_id);
-        session->book_id = NULL;
-        qof_session_push_error (session, err, msg);
-        /* qof_backend_get_message transfers ownership. */
-	g_free(msg);
+        m_book_id = {};
+        push_error (err, msg ? msg : "");
     }
+    g_free (msg);
 }
 
-
-/* ====================================================================== */
-gboolean
-qof_session_save_in_progress(const QofSession *session)
+void
+QofSessionImpl::ensure_all_data_loaded () noexcept
 {
-    return (session && g_atomic_int_get(&session->lock) != 1);
+    if (!m_backend) return;
+    if (!m_backend->load) return;
+    m_backend->load(m_backend, get_book (), LOAD_TYPE_LOAD_ALL);
+    push_error (qof_backend_get_error (m_backend), {});
 }
 
 void
-qof_session_end (QofSession *session)
+QofSessionImpl::swap_books (QofSessionImpl & other) noexcept
+{
+    ENTER ("sess1=%p sess2=%p", this, &other);
+    // don't swap (that is, double-swap) read_only flags
+    std::swap (m_book->read_only, other.m_book->read_only);
+    std::swap (m_book, other.m_book);
+    qof_book_set_backend (m_book, other.m_backend);
+    qof_book_set_backend (other.m_book, m_backend);
```

Comment:
The books already have the right backend associated with them. It's m_backend that needs to be changed.